### PR TITLE
feat: move config to parameter instead of global

### PR DESCRIPTION
**Pull request by fzipi**
Signed-off-by: Felipe Zipitria <felipe.zipitria@owasp.org>

Removes global config. Now some methods will receive the extra config parameter.

Partially addresses #108.

### CODE REVIEW

**Review by anuraaga — DISMISSED** [review]
Nice!

File: config/config.go [hunk 1]
```diff
@@ -42,83 +62,76 @@ func NewConfigFromFile(cfgFile string) error {
...
-	loadDefaults()
 
-	return err
+	return cfg, err
 }
 
 // NewConfigFromEnv reads configuration information from environment variables that start with `FTW_`
-func NewConfigFromEnv() error {
+func NewConfigFromEnv() (*FTWConfiguration, error) {
 	// koanf merges by default, but we never want to merge in this case
```
Comment (theseion) [inline config/config.go:85]:
```suggestion
```

File: config/config.go [hunk 2]
```diff
@@ -42,83 +62,76 @@ func NewConfigFromFile(cfgFile string) error {
...
+	err = k.UnmarshalWithConf("", cfg, koanf.UnmarshalConf{Tag: "koanf"})
 
-	return err
+	return cfg, err
 }
 
 // NewConfigFromString initializes the configuration from a yaml formatted string. Useful for testing.
-func NewConfigFromString(conf string) error {
+func NewConfigFromString(conf string) (*FTWConfiguration, error) {
 	// koanf merges by default, but we never want to merge in this case
```
Comment (theseion) [inline config/config.go:106]:
```suggestion
```

File: config/config.go [hunk 3]
```diff
@@ -42,83 +62,76 @@ func NewConfigFromFile(cfgFile string) error {
...
-	// will overwrite defaults when the associated field is empty in nested
-	// structures (top level would work). That's why we set defaults here
-	// explicitly.
-	if FTWConfig.LogMarkerHeaderName == "" {
-		FTWConfig.LogMarkerHeaderName = DefaultLogMarkerHeaderName
-	}
-	if FTWConfig.RunMode == "" {
-		FTWConfig.RunMode = DefaultRunMode
-	}
+func (c *FTWConfiguration) WithOverrides(overrides FTWTestOverride) {
```
Comment (theseion) [inline config/config.go:127]: Add method comment

File: config/config.go [hunk 4]
```diff
@@ -42,83 +62,76 @@ func NewConfigFromFile(cfgFile string) error {
...
+	c.TestOverride = overrides
+}
+
+func (c *FTWConfiguration) WithRunMode(runmode RunMode) {
```
Comment (theseion) [inline config/config.go:131]: Add method comment

File: config/config.go [hunk 5]
```diff
@@ -42,83 +62,76 @@ func NewConfigFromFile(cfgFile string) error {
...
+	c.RunMode = runmode
+}
 
+func (c *FTWConfiguration) WithLogMarkerHeaderName(name string) {
```
Comment (theseion) [inline config/config.go:135]: Add method comment.

File: config/config_test.go [hunk 6]
```diff
@@ -75,72 +100,96 @@ func TestNewConfigDefaultConfig(t *testing.T) {
...
+	assert.NotNil(t, cfg)
+	assert.Equalf(t, CloudRunMode, cfg.RunMode,
+		"unexpected value '%s' for run mode, expected '%s;", cfg.RunMode, CloudRunMode)
+}
 
-	assert.Equalf(t, CloudRunMode, FTWConfig.RunMode,
-		"unexpected value '%s' for run mode, expected '%s;", FTWConfig.RunMode, CloudRunMode)
+func TestNewDefaultConfigWithParams(t *testing.T) {
+	cfg := NewDefaultConfig()
+	assert.Equal(t, DefaultLogMarkerHeaderName, cfg.LogMarkerHeaderName)
```
Comment (theseion) [inline config/config_test.go:178]: You have a separate test for the default configuration. I wouldn't repeat the assertions from that test here.

File: runner/run.go [hunk 7]
```diff
@@ -20,13 +21,13 @@ import (
 
 var errBadTestRequest = errors.New("ftw/run: bad test: choose between data, encoded_request, or raw_request")
 
-// Run runs your tests with the specified Config.
-func Run(tests []test.FTWTest, c Config, out *output.Output) (*TestRunContext, error) {
+// Run runs your tests with the specified Config. Returns error if some test failed
```
Comment (theseion) [inline runner/run.go:24]: That part of the comment was removed on purpose.
```suggestion
// Run runs your tests with the specified Config.
```

File: runner/run.go [hunk 8]
```diff
@@ -38,7 +39,7 @@ func Run(tests []test.FTWTest, c Config, out *output.Output) (*TestRunContext, e
 	}
 	client, err := ftwhttp.NewClient(conf)
 	if err != nil {
-		return &TestRunContext{}, err
+		return TestRunContext{}, err
```
Comment (theseion) [inline runner/run.go:42]: Out of curiosity: why did you change from pointer to value for the error?

File: waflog/waflog_test.go [hunk 9]
```diff
@@ -9,24 +9,17 @@ import (
...
 
 func TestNewFTWLogLines(t *testing.T) {
-	err := config.NewConfigFromEnv()
-	assert.NoError(t, err)
+	cfg := config.NewDefaultConfig()
+	assert.NotNil(t, cfg)
 
 	// Don't call NewFTWLogLines to avoid opening the file.
 	ll := &FTWLogLines{}
 	// Loop through each option
```
Comment (theseion) [inline waflog/waflog_test.go:17]:
```suggestion
```

**Review by theseion — CHANGES_REQUESTED** [review]

**Comment by theseion** [comment]
Awesome!

File: config/config.go (hunk 3 above)
Comment (fzipi) [reply config/config.go:125]: Done.

File: config/config.go (hunk 5 above)
Comment (fzipi) [reply config/config.go:133]: Added.

File: config/config_test.go (hunk 6 above)
Comment (fzipi) [reply config/config_test.go:178]: Removed.

File: runner/run.go (hunk 8 above)
Comment (fzipi) [reply runner/run.go:42]: I don't remember, but let's revert the change until the reason is clear 😄

File: config/config.go (hunk 4 above)
Comment (fzipi) [reply config/config.go:129]: Added.

**Review by theseion — APPROVED** [review]
